src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs

```diff
@@ -108,22 +124,38 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             // If what we have here is known to be a type, we don't want to clutter the variable name suggestion list
             // with 'when' since we know that the resulting code would be semantically invalid.
 
-            var expression = expressionOrPattern as ExpressionSyntax
-                ?? (expressionOrPattern as ConstantPatternSyntax)?.Expression;
-            
-            if (!(expression is TypeSyntax typeSyntax))
+            bool isVar;
+            ImmutableArray<ISymbol> symbols;
+
+            if (nodeOrToken.IsNode)
```

Comment:
i think it woudl be worth documenting which cases each part of the if is trying to handle.